src/BatchRestore/Program.cs

```diff
@@ -0,0 +1,147 @@
+    /// of instances of dotnet-restore we need to launch we batch up folders to pass to each invocation, being
+    /// mindful of the command line limits of Windows.
+    /// 
+    /// This program needs to be placed next to `dotnet-restore` and can then be launched with 
+    /// `dotnet batch-restore`.  Note we use CoreFX command line argument styles (i.e. -foo instead of
+    /// --foo).
+    /// </summary>
+    internal class Program
+    {
+        private static List<string> s_directoriesToRestore = new List<string>();
+        private const int CommandLineLengthLimit = 7500; /* Well under the 8191 limit of Windows */
+
+        private static int Main(string[] args)
+        {
+            ParseCommandLine(args);
+
+            foreach (string directory in s_rootDirectories)
+            {
+                VisitDirectory(directory, true);
+            }
+
+            string argumentPrefix = GetDotnetRestoreCommandPrefix();
+
+            // Loop over all the directories to restore, invoking dotnet-restore as few times as possible
+            // by batching things together.
+            StringBuilder argumentBuilder = new StringBuilder(argumentPrefix, CommandLineLengthLimit);
+            foreach(string directoryToRestore in s_directoriesToRestore)
+            {
+                argumentBuilder.Append(" \"");
+                argumentBuilder.Append(directoryToRestore);
+                argumentBuilder.Append('"');
+
+                if (argumentBuilder.Length > CommandLineLengthLimit)
+                {
+                    int retValue = RunDotnetRestore(argumentBuilder);
+
+                    if (retValue != 0)
+                    {
+                        return retValue;
+                    }
+
+                    argumentBuilder.Clear();
+                    argumentBuilder.Append(argumentPrefix);
+                }
+            }
+
+            if (argumentBuilder.Length > argumentPrefix.Length)
+            {
+                return RunDotnetRestore(argumentBuilder);
+            }
+
+            return 0;
+        }
+
+        private static int RunDotnetRestore(StringBuilder argumentBuilder)
+        {
+            string DotnetCommandName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "dotnet.exe" : "dotnet";
+
+            ProcessStartInfo ps = new ProcessStartInfo();
+            ps.CreateNoWindow = false;
+            ps.UseShellExecute = false;
+            ps.FileName = Path.Combine(AppContext.BaseDirectory, DotnetCommandName);
+            ps.Arguments = argumentBuilder.ToString();
+
+            using (Process p = Process.Start(ps))
+            {
+                p.WaitForExit();
+                return p.ExitCode;
+            }
+        }
+
+        private static void VisitDirectory(string directory, bool shouldAddDirectoryToRestoreList)
+        {
+            if (shouldAddDirectoryToRestoreList)
+            {
+                s_directoriesToRestore.Add(directory);
```

Comment:
Does this add *all* subdirs to the command where a project.json is in a dir? Or is there some filter I'm not seeing for directories without any project.jsons in them? (Since those directories don't need to be restored.) I don't think it makes a difference for dnu restore time, but it might let us fit more on a single command. It might not be worth the extra complexity though.